d2renderers/d2svg/d2svg.go

```diff
@@ -494,6 +495,191 @@ func makeLabelMask(labelTL *geo.Point, width, height int, opacity float64) strin
+	u0, u1 := 1-t0, 1-t1
+
+	q1 := geo.Point{
+		X: (u0*u0*u0)*p1.X + (3*t0*u0*u0)*p2.X + (3*t0*t0*u0)*p3.X + t0*t0*t0*p4.X,
+		Y: (u0*u0*u0)*p1.Y + (3*t0*u0*u0)*p2.Y + (3*t0*t0*u0)*p3.Y + t0*t0*t0*p4.Y,
+	}
+	q2 := geo.Point{
+		X: (u0*u0*u1)*p1.X + (2*t0*u0*u1+u0*u0*t1)*p2.X + (t0*t0*u1+2*u0*t0*t1)*p3.X + t0*t0*t1*p4.X,
+		Y: (u0*u0*u1)*p1.Y + (2*t0*u0*u1+u0*u0*t1)*p2.Y + (t0*t0*u1+2*u0*t0*t1)*p3.Y + t0*t0*t1*p4.Y,
+	}
+	q3 := geo.Point{
+		X: (u0*u1*u1)*p1.X + (t0*u1*u1+2*u0*t1*u1)*p2.X + (2*t0*t1*u1+u0*t1*t1)*p3.X + t0*t1*t1*p4.X,
+		Y: (u0*u1*u1)*p1.Y + (t0*u1*u1+2*u0*t1*u1)*p2.Y + (2*t0*t1*u1+u0*t1*t1)*p3.Y + t0*t1*t1*p4.Y,
+	}
+	q4 := geo.Point{
+		X: (u1*u1*u1)*p1.X + (3*t1*u1*u1)*p2.X + (3*t1*t1*u1)*p3.X + t1*t1*t1*p4.X,
+		Y: (u1*u1*u1)*p1.Y + (3*t1*u1*u1)*p2.Y + (3*t1*t1*u1)*p3.Y + t1*t1*t1*p4.Y,
+	}
+
+	return q1, q2, q3, q4
+}
+
+// Adds a certain line/curve to an existing SVG path string. offsetIdx and pathData provides the points needed
+func addToPath(path *string, pathType *string, offsetIdx int, pathData []string) int {
+	var increment int
+
+	switch *pathType {
+	case "M":
+		*path += fmt.Sprintf("M %s %s ", pathData[offsetIdx+1], pathData[offsetIdx+2])
+		increment = 3
+	case "L":
+		*path += fmt.Sprintf("L %s %s ", pathData[offsetIdx+1], pathData[offsetIdx+2])
+		increment = 3
+	case "C":
+		*path += fmt.Sprintf("C %s %s %s %s %s %s ", pathData[offsetIdx+1], pathData[offsetIdx+2], pathData[offsetIdx+3], pathData[offsetIdx+4], pathData[offsetIdx+5], pathData[offsetIdx+6])
+		increment = 7
+	case "S":
+		*path += fmt.Sprintf("S %s %s %s %s ", pathData[offsetIdx+1], pathData[offsetIdx+2], pathData[offsetIdx+3], pathData[offsetIdx+4])
+		increment = 5
+	default:
+		panic(fmt.Sprintf("unknown svg path command \"%s\"", pathData[offsetIdx]))
+	}
+
+	return increment
+}
+
+// This function finds the length of a path in SVG notation
+func pathLength(pathData []string) float64 {
+	var x, y, pathLength float64
+	var increment int
+	var prevPosition geo.Point
+
+	for i := 0; i < len(pathData); {
+		switch pathData[i] {
+		case "M":
+			x, _ = strconv.ParseFloat(pathData[i+1], 64)
+			y, _ = strconv.ParseFloat(pathData[i+2], 64)
+
+			increment = 3
+		case "L":
+			x, _ = strconv.ParseFloat(pathData[i+1], 64)
+			y, _ = strconv.ParseFloat(pathData[i+2], 64)
+
+			pathLength += geo.EuclideanDistance(prevPosition.X, prevPosition.Y, x, y)
+
+			increment = 3
+		case "C":
+			x, _ = strconv.ParseFloat(pathData[i+5], 64)
+			y, _ = strconv.ParseFloat(pathData[i+6], 64)
+
+			pathLength += geo.EuclideanDistance(prevPosition.X, prevPosition.Y, x, y)
+
+			increment = 7
+		case "S":
+			x, _ = strconv.ParseFloat(pathData[i+3], 64)
+			y, _ = strconv.ParseFloat(pathData[i+4], 64)
+
+			pathLength += geo.EuclideanDistance(prevPosition.X, prevPosition.Y, x, y)
+
+			increment = 5
+		default:
+			panic(fmt.Sprintf("unknown svg path command \"%s\"", pathData[i]))
+		}
+
+		prevPosition = geo.Point{X: x, Y: y}
+		i += increment
```

Comment:
why's this not just in the for loop?

Reply:
Sorry, I'm not really sure what you mean by this. What should be in the for loop?

Reply:
Above your for loop is `for i := 0; i < len(pathData); {`

Why not `for i := 0; i < len(pathData); i += increment {` which is a standard 3-part for loop

Reply:
Got it, makes sense.